rclpy/src/rclpy/context.cpp

```diff
@@ -30,93 +30,93 @@
 
 namespace rclpy
 {
-size_t
-context_get_domain_id(py::capsule pycontext)
+Context::Context()
 {
-  auto context = static_cast<rcl_context_t *>(
-    rclpy_handle_get_pointer_from_capsule(pycontext.ptr(), "rcl_context_t"));
-  if (!context) {
-    throw py::error_already_set();
-  }
+  // Create a client
+  rcl_context_ = std::shared_ptr<rcl_context_t>(
+    new rcl_context_t,
+    [](rcl_context_t * context)
+    {
+      if (NULL != context->impl) {
+        rcl_ret_t ret;
+        if (rcl_context_is_valid(context)) {
+          // shutdown first, if still valid
+          ret = rcl_shutdown(context);
+          if (RCL_RET_OK != ret) {
+            RCUTILS_SAFE_FWRITE_TO_STDERR(
+              "[rclpy|" RCUTILS_STRINGIFY(__FILE__) ":" RCUTILS_STRINGIFY(__LINE__) "]: "
+              "failed to shutdown rcl_context_t");
+            RCUTILS_SAFE_FWRITE_TO_STDERR(rcl_get_error_string().str);
+            RCUTILS_SAFE_FWRITE_TO_STDERR("\n");
+            rcl_reset_error();
+          }
+        }
+        ret = rcl_context_fini(context);
+        if (RCL_RET_OK != ret) {
+          RCUTILS_SAFE_FWRITE_TO_STDERR(
+            "[rclpy|" RCUTILS_STRINGIFY(__FILE__) ":" RCUTILS_STRINGIFY(__LINE__) "]: "
+            "failed to fini rcl_context_t");
+          RCUTILS_SAFE_FWRITE_TO_STDERR(rcl_get_error_string().str);
+          RCUTILS_SAFE_FWRITE_TO_STDERR("\n");
+          rcl_reset_error();
+        }
+      }
+      delete context;
+    });
+  *rcl_context_ = rcl_get_zero_initialized_context();
+}
 
+void
+Context::destroy()
+{
+  rcl_context_.reset();
```

Comment:
@ahcorde I see the test you mention `SIGABRT`'ing. I think it's because we're not ensuring destruction order. I see `Context` dependent objects keeping shared ownership of that `Context`, but that won't prevent `rcl_context_t` destruction if something calls `Context::destroy()` on it.

Reply:
This seems like a design problem.

`rcl_context_t` needs to outlive `rcl_node_t`
`rcl_node_t` needs to outlive `rcl_publisher_t`.

`rclpy::Publisher` keeps `rcl_publisher_t` alive, until destructed or `destroy()`
`rclpy::Publisher` keeps `rclpy::Node` alive, until destructed or `destroy()`

`rclpy::Node` keeps `rcl_node_t` alive, until destructed or `destroy()`
`rclpy::Node` keeps `rclpy::Context` alive, until destructed or `destroy()`

`rclpy::Context` keeps `rcl_context_t` alive, until destructed or `destroy()`


`rclpy::Publisher` holding `std::shared_ptr<rclpy::Node>` is assuming keeping `rclpy::Node` alive will keep `rcl_node_t` and `rcl_context_t` alive, but that isn't true.
`rclpy::Node` holding `std::shared_ptr<rclpy::Context>` is assuming keeping `rclpy::Context` alive will keep `rcl_context_t` alive, but that isn't true.

How do we fix this?

Reply:
Hmm, to replicate the destruction semantics we have today, having each wrapper object hold shared ownership of its parents' underlying C structs seems like the only path forward. It's not ideal, and somewhat defeats the purpose of early destruction as long as you have some dependent object in the garbage collector hands, but that's what `rclpy_handle_t` did.

Alternatively, we could have each wrapper object hold weak ownership of its children, so as to propagate an early destruction request down the dependency tree. Thoughts?

Reply:
> to replicate the destruction semantics we have today, having each wrapper object hold shared ownership of its parents' underlying C structs seems like a path forward.

This seems like a reasonable path forward. Children need some way to keep the `rcl` part of their parent's alive. One way is to put those `rcl` pointers into their own structs.

<details><summary>Psuedocode of holding parent's C structs</summary>

```C++
class rclpy::Node
{
  struct RclPtrs
  {
    std::shared_ptr<rclpy::Context::RclPtrs> context_ptrs;
    std::unique_ptr<rcl_node_t> rcl_node;
  };

  /// Return RCL pointers so another class can keep the rcl part alive
  std::shared_ptr<rclpy::Node::RclPtrs>
  get_rcl_ptrs() const
  {
    return rcl_ptrs_;
  }

  void
  destroy()
  {
    // Drop references here, and rely on Python `class Node` to destroy children so all shared references to struct are dropped
    rcl_ptrs_.reset();
  }

  std::shared_ptr<rclpy::Node::RclPtrs> rcl_ptrs_;
}


class rclpy::Publisher
{
  struct RclPtrs
  {
    // Store node_ptrs in here so rcl_event_t can keep node ptrs alive too
    std::shared_ptr<rclpy::Node::RclPtrs> node_ptrs;
    std::unique_ptr<rcl_publisher_t> rcl_publisher;
  };

  Publisher(const rclpy::Node & node)
  : rcl_ptrs_(new rclpy::Publisher::RclPtrs)
  {
    rcl_ptrs_->node_ptrs = node->get_rcl_ptrs();
    /// ...
  }

  void
  destroy()
  {
    rcl_ptrs_.reset();
  }

  std::shared_ptr<rclpy::Publisher::RclPtrs> rcl_ptrs_;
}
```
</details>

The downside of this way is the publisher instance no longer has access to the Node's API. Maybe the publisher can have both if it intentionally makes a copy of  `rclpy::Node`?

<details><summary>Psuedocode of intentional node copy</summary>

```c++
class rclpy::Node
{
  // Copy constructor
  Node(const Node & other)
  : rcl_context_(other.rcl_context_), rcl_node_(other.rcl_node_)
  {
    // Gives copy shared ownership of the rcl pointers
  }

  void
  destroy()
  {
    // Drop references if we can, and rely on Python `class Node` to destroy children so all copies are destroyed too
    rcl_ptrs_.reset();
  }

  // Shared pointers so ownership can be shared with copies
  std::shared_ptr<rcl_context_t> rcl_context_;
  std::shared_ptr<rcl_node_t> rcl_node_;
}


class rclpy::Publisher
{
  Publisher(const rclpy::Node & node)
  : node_(node)  // Make a copy
  {
  }

  void
  destroy()
  {
    rcl_publisher_.reset();
    node_.destroy();
  }

  // Publisher has both access to the `rclpy::Node` API, and ownership of the rcl pointers
  rclpy::Node node_;
  std::shared_ptr<rcl_publisher_t> rcl_publisher_;
}
```
</details>

> It's not ideal, and somewhat defeats the purpose of early destruction as long as you have some dependent object in the garbage collector hands, but that's what rclpy_handle_t did.

I think the expectation with `rclpy_handle_t` was the Python object, `class Node`, took care of that in `destroy_node()` by calling `destroy()` on all of it's children first. I think that expectation can stay the same.

> Alternatively, we could have each wrapper object hold weak ownership of its children, so as to propagate an early destruction request down the dependency tree. Thoughts?

Say `Node` has a weak pointer to `Publisher`, and a user calls `Node::destroy()`. `Node` doesn't call `Publisher::destroy()`, because what if someone is still using `Publisher`? If `Node` calls `Publisher::destroy_when_not_in_use()` and `Publisher` is busy, then what? Does the publisher use its reference to the `Node` so that it can call `Node::destroy()` again when the `Publisher` is destroyed?

Node` having a weak reference to `Publisher` means `Node` must be told when the publisher depends on it. Maybe that would be `Publisher` calling an API on `Node` to say it was constructed?

I wish `rcl` handled destruction order - maybe by blocking `rcl_node_t` from being fini'd as long as another `rcl_*_t` was using it.

Reply:
I implemented the first solution that @sloretz mentioned in this commit https://github.com/ros2/rclpy/pull/771/commits/7b0c4dcbb4a6d8f0904087004a92a99221c3754e

But I still have some problems, in this case with `test_publisher`, this is the backtrace:

```
Thread 1 "python3" received signal SIGSEGV, Segmentation fault.
0x00007fffce1b1b2f in std::enable_if<std::__and_<std::__not_<std::__is_tuple_like<rcl_publisher_t*> >, std::is_move_constructible<rcl_publisher_t*>, std::is_move_assignable<rcl_publisher_t*> >::value, void>::type std::swap<rcl_publisher_t*>(rcl_publisher_t*&, rcl_publisher_t*&) () from /home/ahcorde/ros2_galactic/build/rclpy/test_rclpy/_rclpy_pybind11.cpython-38-x86_64-linux-gnu.so
(gdb) bt
#0  0x00007fffce1b1b2f in std::enable_if<std::__and_<std::__not_<std::__is_tuple_like<rcl_publisher_t*> >, std::is_move_constructible<rcl_publisher_t*>, std::is_move_assignable<rcl_publisher_t*> >::value, void>::type std::swap<rcl_publisher_t*>(rcl_publisher_t*&, rcl_publisher_t*&) () from /home/ahcorde/ros2_galactic/build/rclpy/test_rclpy/_rclpy_pybind11.cpython-38-x86_64-linux-gnu.so
#1  0x00007fffce1b09e3 in std::unique_ptr<rcl_publisher_t, std::function<void (rcl_publisher_t*)> >::reset(rcl_publisher_t*) ()
   from /home/ahcorde/ros2_galactic/build/rclpy/test_rclpy/_rclpy_pybind11.cpython-38-x86_64-linux-gnu.so
#2  0x00007fffce1afd21 in rclpy::Publisher::destroy() () from /home/ahcorde/ros2_galactic/build/rclpy/test_rclpy/_rclpy_pybind11.cpython-38-x86_64-linux-gnu.so
#3  0x00007fffce18a5fc in rclpy::Destroyable::destroy_when_not_in_use() () from /home/ahcorde/ros2_galactic/build/rclpy/test_rclpy/_rclpy_pybind11.cpython-38-x86_64-linux-gnu.so
#4  0x00007fffce18b3e3 in pybind11::cpp_function::cpp_function<void, rclpy::Destroyable, , pybind11::name, pybind11::is_method, pybind11::sibling, char [74]>(void (rclpy::Destroyable::*)(), pybind11::name const&, pybind11::is_method const&, pybind11::sibling const&, char const (&) [74])::{lambda(rclpy::Destroyable*)#1}::operator()(rclpy::Destroyable*) const ()
   from /home/ahcorde/ros2_galactic/build/rclpy/test_rclpy/_rclpy_pybind11.cpython-38-x86_64-linux-gnu.so
#5  0x00007fffce18d105 in void pybind11::detail::argument_loader<rclpy::Destroyable*>::call_impl<void, pybind11::cpp_function::cpp_function<void, rclpy::Destroyable, , pybind11::name, pybind11::is_method, pybind11::sibling, char [74]>(void (rclpy::Destroyable::*)(), pybind11::name const&, pybind11::is_method const&, pybind11::sibling const&, char const (&) [74])::{lambda(rclpy::Destroyable*)#1}&, 0ul, pybind11::detail::void_type>(pybind11::cpp_function::cpp_function<void, rclpy::Destroyable, , pybind11::name, pybind11::is_method, pybind11::sibling, char [74]>(void (rclpy::Destroyable::*)(), pybind11::name const&, pybind11::is_method const&, pybind11::sibling const&, char const (&) [74])::{lambda(rclpy::Destroyable*)#1}&, std::integer_sequence<unsigned long, 0ul>, pybind11::detail::void_type&&) && ()
   from /home/ahcorde/ros2_galactic/build/rclpy/test_rclpy/_rclpy_pybind11.cpython-38-x86_64-linux-gnu.so
#6  0x00007fffce18c2fe in std::enable_if<std::is_void<void>::value, pybind11::detail::void_type>::type pybind11::detail::argument_loader<rclpy::Destroyable*>::call<void, pybind11::detail::void_type, pybind11::cpp_function::cpp_function<void, rclpy::Destroyable, , pybind11::name, pybind11::is_method, pybind11::sibling, char [74]>(void (rclpy::Destroyable::*)(), pybind11::name const&, pybind11::is_method const&, pybind11::sibling const&, char const (&) [74])::{lambda(rclpy::Destroyable*)#1}&>(pybind11::cpp_function::cpp_function<void, rclpy::Destroyable, , pybind11::name, pybind11::is_method, pybind11::sibling, char [74]>(void (rclpy::Destroyable::*)(), pybind11::name const&, pybind11::is_method const&, pybind11::sibling const&, char const (&) [74])::{lambda(rclpy::Destroyable*)#1}&) && ()
   from /home/ahcorde/ros2_galactic/build/rclpy/test_rclpy/_rclpy_pybind11.cpython-38-x86_64-linux-gnu.so
#7  0x00007fffce18bd79 in pybind11::cpp_function::initialize<pybind11::cpp_function::initialize<void, rclpy::Destroyable, , pybind11::name, pybind11::is_method, pybind11::sibling, char [74]>(void (rclpy::Destroyable::*)(), pybind11::name const&, pybind11::is_method const&, pybind11::sibling const&, char const (&) [74])::{lambda(rclpy::Destroyable*)#1}, void, rclpy::Destroyable*, pybind11::name, pybind11::is_method, pybind11::sibling, char [74]>(pybind11::cpp_function::initialize<void, rclpy::Destroyable, , pybind11::name, pybind11::is_method, pybind11::sibling, char [74]>(void (rclpy::Destroyable::*)(), pybind11::name const&, pybind11::is_method const&, pybind11::sibling const&, char const (&) [74])::{lambda(rclpy::Destroyable*)#1}&&, void (*)(rclpy::Destroyable*), pybind11::name const&, pybind11::is_method const&, pybind11::sibling const&, char const (&) [74])::{lambda(pybind11::detail::function_call&)#3}::operator()(pybind11::detail::function_call) const ()
   from /home/ahcorde/ros2_galactic/build/rclpy/test_rclpy/_rclpy_pybind11.cpython-38-x86_64-linux-gnu.so
#8  0x00007fffce18bde1 in pybind11::cpp_function::initialize<pybind11::cpp_function::initialize<void, rclpy::Destroyable, , pybind11::name, pybind11::is_method, pybind11::sibling, char [74]>(void (rclpy::Destroyable::*)(), pybind11::name const&, pybind11::is_method const&, pybind11::sibling const&, char const (&) [74])::{lambda(rclpy::Destroyable*)#1}, void, rclpy::Destroyable*, pybind11::name, pybind11::is_method, pybind11::sibling, char [74]>(pybind11::cpp_function::initialize<void, rclpy::Destroyable, , pybind11::name, pybind11::is_method, pybind11::sibling, char [74]>(void (rclpy::Destroyable::*)(), pybind11::name const&, pybind11::is_method const&, pybind11::sibling const&, char const (&) [74])::{lambda(rclpy::Destroyable*)#1}&&, void (*)(rclpy::Destroyable*), pybind11::name const&, pybind11::is_method const&, pybind11::sibling const&, char const (&) [74])::{lambda(pybind11::detail::function_call&)#3}::_FUN(pybind11::detail::function_call) ()
   from /home/ahcorde/ros2_galactic/build/rclpy/test_rclpy/_rclpy_pybind11.cpython-38-x86_64-linux-gnu.so
#9  0x00007fffce0f4095 in pybind11::cpp_function::dispatcher(_object*, _object*, _object*) () from /home/ahcorde/ros2_galactic/build/rclpy/test_rclpy/_rclpy_pybind11.cpython-38-x86_64-linux-gnu.so
#10 0x00000000005f2fb9 in cfunction_call_varargs (kwargs=<optimized out>, args=<optimized out>, func=<built-in method destroy_when_not_in_use of PyCapsule object at remote 0x7fffce2c4b40>)
    at ../Objects/call.c:772
#11 PyCFunction_Call (func=<built-in method destroy_when_not_in_use of PyCapsule object at remote 0x7fffce2c4b40>, args=<optimized out>, kwargs=<optimized out>) at ../Objects/call.c:772
#12 0x00000000005f3446 in _PyObject_MakeTpCall (callable=<built-in method destroy_when_not_in_use of PyCapsule object at remote 0x7fffce2c4b40>, args=<optimized out>, nargs=<optimized out>, 
    keywords=<optimized out>) at ../Include/internal/pycore_pyerrors.h:13
#13 0x000000000050aa8b in _PyObject_Vectorcall (kwnames=0x0, nargsf=<optimized out>, args=0x7fffcc06e6d0, callable=<built-in method destroy_when_not_in_use of PyCapsule object at remote 0x7fffce2c4b40>)
    at ../Include/cpython/abstract.h:125
#14 _PyObject_Vectorcall (kwnames=0x0, nargsf=<optimized out>, args=0x7fffcc06e6d0, callable=<built-in method destroy_when_not_in_use of PyCapsule object at remote 0x7fffce2c4b40>)
    at ../Include/cpython/abstract.h:115
#15 method_vectorcall (method=<optimized out>, args=0x7fffcc06e6d8, nargsf=<optimized out>, kwnames=0x0) at ../Objects/classobject.c:60
#16 0x000000000056ef97 in _PyObject_Vectorcall (kwnames=0x0, nargsf=<optimized out>, args=0x7fffcc06e6d8, callable=<method at remote 0x7ffff5b75180>) at ../Include/cpython/abstract.h:127
#17 call_function (kwnames=0x0, oparg=<optimized out>, pp_stack=<synthetic pointer>, tstate=0x962f70) at ../Python/ceval.c:4963
#18 _PyEval_EvalFrameDefault (f=<optimized out>, throwflag=<optimized out>) at ../Python/ceval.c:3469
#19 0x00000000005f5e56 in PyEval_EvalFrameEx (throwflag=0, 
    f=Frame 0x7fffcc06e550, for file /home/ahcorde/ros2_galactic/src/ros2/rclpy/rclpy/rclpy/publisher.py, line 92, in destroy (self=<Publisher(_Publisher__publisher=<test_rclpy._rclpy_pybind11.Publisher at remote 0x7fffce01dfb0>, msg_type=<Metaclass_BasicTypes(__module__='test_msgs.msg._basic_types', __doc__="Message class 'BasicTypes'.", __slots__=['_bool_value', '_byte_value', '_char_value', '_float32_value', '_float64_value', '_int8_value', '_uint8_value', '_int16_value', '_uint16_value', '_int32_value', '_uint32_value', '_int64_value', '_uint64_value'], _fields_and_field_types={'bool_value': 'boolean', 'byte_value': 'octet', 'char_value': 'uint8', 'float32_value': 'float', 'float64_value': 'double', 'int8_value': 'int8', 'uint8_value': 'uint8', 'int16_value': 'int16', 'uint16_value': 'uint16', 'int32_value': 'int32', 'uint32_value': 'uint32', 'int64_value': 'int64', 'uint64_value': 'uint64'}, SLOT_TYPES=(<BasicType at remote 0x7fffce344790>, <BasicType at remote 0x7fffce3447c0>, <BasicType at remote 0x7fffce344...(truncated))
    at ../Python/ceval.c:741
#20 function_code_fastcall (globals=<optimized out>, nargs=<optimized out>, args=<optimized out>, co=<optimized out>) at ../Objects/call.c:283
#21 _PyFunction_Vectorcall (func=<optimized out>, stack=<optimized out>, nargsf=<optimized out>, kwnames=<optimized out>) at ../Objects/call.c:410
#22 0x000000000056a136 in _PyObject_Vectorcall (kwnames=0x0, nargsf=<optimized out>, args=0x7fffcc069b28, callable=<function at remote 0x7fffcc2edb80>) at ../Include/cpython/abstract.h:127
```

Reply:
>  If `Node` calls `Publisher::destroy_when_not_in_use()` and `Publisher` is busy, then what?

Then neither gets destroyed, but both are "scheduled" for destruction, blocking further use cases. Then, if you have dependent objects increasing/decreasing their dependencies' use count in topological order when theirs is (i.e. on `__enter__` and on `__exit__`), then they'll get destroyed once all ongoing tasks are complete.

> `Node` having a weak reference to `Publisher` means `Node` must be told when the `Publisher` depends on it. 

Indeed. But that's something we can do internally, when the `Publisher` is constructed through a non-binded API. We can also use friendship if we want to hide that API in C++.

Reply:
@ahcorde I pushed a few commits to this branch that get all tests except cppcheck to pass, but it's not quite complete. The ActionClient, Client, Timer, etc classes need to use the RclPtrs structs of their parents too or they won't keep the `rcl` part of them alive.

895ee4b prevents a Destroyable from being destroyed twice. This was happening because of the publisher being explicitly destroyed in `test_publisher`, and `Node.__del__` was trying to destroy the publisher again. I removed the `Node.__del__` in 21e80db. It shouldn't be needed if everything else is working correctly. The remaining bugs were fixed by reordering the pointers in the `RclPtrs` structs in ad53180. C++ destroys structs in reverse order, so pointers to parents need to be placed above children created from them.

After looking at it I think I like the intentional copy approach more. I'll try prototyping that and open a PR to this branch if it works.

Reply:
@sloretz I prototyped the second option in this branch https://github.com/ros2/rclpy/pull/new/ahcorde/pybind11-node_option2, 

I cherry-picked two of your improvements but I'm having some issues with qos_event tests

Reply:
@ahcorde I opened https://github.com/ros2/rclpy/pull/773 which has a few differences, and all tests pass. Mind taking a look?

The qos_event tests are written in a way that makes them flaky when we do destruction ordering correctly, but I included a fix on my branch.